tests/test_resolver.py

```diff
@@ -1,157 +1,281 @@
+        Resolver().workspace_from_url(None)
+
+    # check exception
+    assert "Must pass 'mets_url'" in str(exc)
+
+
+@mock.patch("requests.get")
+def test_workspace_from_url_kant(mock_request, tmp_path):
+
+    # arrange
+    url_src = 'https://raw.githubusercontent.com/OCR-D/assets/master/data/kant_aufklaerung_1784/data/mets.xml'
+    mock_request.side_effect = request_behavior
+    dst_dir = tmp_path / 'workspace_kant'
+    dst_dir.mkdir()
+
+    # act
+    resolver = Resolver()
+    resolver.workspace_from_url(
+        url_src, mets_basename='foo.xml', dst_dir=dst_dir)
+
+    # assert
+    local_path = dst_dir / 'foo.xml'
+    assert os.path.isfile(str(local_path))
+    # 1 time data was requested
+    assert mock_request.call_count == 1
+
+
+@mock.patch("requests.get")
+def test_workspace_from_url_kant_with_resources(mock_request, tmp_path):
+
+    # arrange
+    url_src = 'https://raw.githubusercontent.com/OCR-D/assets/master/data/kant_aufklaerung_1784/data/mets.xml'
+    mock_request.side_effect = request_behavior
+    dst_dir = tmp_path / 'workspace_kant'
+    dst_dir.mkdir()
+
+    # act
+    resolver = Resolver()
+    resolver.workspace_from_url(
+        url_src,
+        mets_basename='kant_aufklaerung_1784.xml',
+        dst_dir=dst_dir,
+        download=True)
+
+    # assert files present under local tmp_path
+    local_path_mets = dst_dir / 'kant_aufklaerung_1784.xml'
+    assert os.path.isfile(str(local_path_mets))
+    local_path_img1 = dst_dir / 'OCR-D-IMG' / 'INPUT_0017.tif'
+    assert os.path.isfile(str(local_path_img1))
+    local_path_page1 = dst_dir / 'OCR-D-GT-PAGE' / 'PAGE_0017_PAGE.xml'
+    assert os.path.isfile(str(local_path_page1))
+
+    # 1 METS/MODS + 2 images + 4 OCR files = 7 requests
+    assert mock_request.call_count == 7
+
+
+@mock.patch("requests.get")
+def test_workspace_from_url_kant_with_resources_existing_local(mock_request, tmp_path):
+
+    # arrange
+    url_src = 'https://raw.githubusercontent.com/OCR-D/assets/master/data/kant_aufklaerung_1784/data/mets.xml'
+    mock_request.side_effect = request_behavior
+    dst_dir = tmp_path / 'workspace_kant'
+    dst_dir.mkdir()
+    src_mets = Path(assets.path_to(
+        'kant_aufklaerung_1784-binarized/data/mets.xml'))
+    dst_mets = Path(dst_dir, 'mets.xml')
+    shutil.copyfile(src_mets, dst_mets)
+
+    # act
+    Resolver().workspace_from_url(url_src,
+                                  clobber_mets=False,
+                                  dst_dir=dst_dir)
+
+    # assert
+    # no real request was made, since mets already present
+    assert mock_request.call_count == 0
+
+
+@mock.patch("requests.get")
+def test_workspace_from_url_404(mock_request):
+    """Expected behavior when try create workspace from invalid online target
+    """
+
+    # arrange
+    url_404 = 'https://raw.githubusercontent.com/OCR-D/assets/master/data/kant_aufklaerung_1784/data/mets.xmlX'
+    mock_request.side_effect = Exception('HTTP request failed')
```

Comment:
Ah, now I get why this is called _side effect_ – it can be much more than just a function!